### PR TITLE
Enable multiparts to be sent and retrieve via http response

### PR DESCRIPTION
**Pull request by arukshani**
Introduce getByteChannel() and setByteChannel() methods to http request and response

Exposed getByteChannel() and setByteChannel() methods that belong to Entity through http request and response 

## Automation tests
 - Unit tests 
   > Added
 - Integration tests
   > Added

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

```ballerina
import ballerina.net.http;
import ballerina.mime;
import ballerina.io;

@http:configuration {port:9093}
service<http> multiparts {
    @http:resourceConfig {
        methods:["GET"],
        path:"/decode_in_response"
    }
    resource multipartReceiver (http:Connection conn, http:InRequest request) {
        endpoint<http:HttpClient> httpEndpoint {
            create http:HttpClient("http://localhost:9092", {});
        }
        http:OutRequest outRequest= {};
        http:InResponse inResponse = {};
        inResponse, _ = httpEndpoint.get("/multiparts/encode_out_response", outRequest);
        mime:Entity[] parentParts = inResponse.getMultiparts();
        int i = 0;
        //Loop through parent parts.
        while (i < lengthof parentParts) {
            mime:Entity parentPart = parentParts[i];
            handleNestedParts(parentPart);
            i = i + 1;
        }
        http:OutResponse res = {};
        res.setStringPayload("Nested Parts Received!");
        _ = conn.respond(res);
    }
}

//Given a parent part, get it's child parts.
function handleNestedParts (mime:Entity parentPart) {
    mime:Entity[] childParts = parentPart.getBodyParts();
    int i = 0;
    if (childParts != null) {
        io:println("Nested Parts Detected!");
        while (i < lengthof childParts) {
            mime:Entity childPart = childParts[i];
            handleContent(childPart);
            i = i + 1;
        }
    } else {
        //When there are no nested parts in a body part, handle the body content directly.
        io:println("Parent doesn't have children. So handling the body content directly...");
        handleContent (parentPart);
    }
}

//Handling body part content logic varies according to user's requirement.
function handleContent (mime:Entity bodyPart) {
    string contentType = bodyPart.contentType.toString();
    if (mime:APPLICATION_XML == contentType || mime:TEXT_XML == contentType) {
        //Extract xml data from body part and print.
        io:println(bodyPart.getXml());
    } else if (mime:APPLICATION_JSON == contentType) {
        //Extract json data from body part and print.
        io:println(bodyPart.getJson());
    } else if (mime:TEXT_PLAIN == contentType){
        //Extract text data from body part and print.
        io:println(bodyPart.getText());
    } else if ("application/vnd.ms-powerpoint" == contentType) {
        //Get a byte channel from body part and write content to a file.
        writeToFile(bodyPart.getByteChannel());
        io:println("Content saved to file");
    }
}

function writeToFile(io:ByteChannel byteChannel) {
    string dstFilePath = "./files/savedFile.ppt";
    io:ByteChannel destinationChannel = getByteChannel(dstFilePath, "w");
    blob readContent;
    int numberOfBytesRead = 1;
    while (numberOfBytesRead != 0) {
        readContent,numberOfBytesRead = byteChannel.readBytes(10000);
        int numberOfBytesWritten = destinationChannel.writeBytes(readContent, 0);
    }
}

function getByteChannel (string filePath, string permission) (io:ByteChannel) {
    io:ByteChannel channel = io:openFile(filePath, permission);
    return channel;
}
```

```ballerina
import ballerina.net.http;
import ballerina.mime;
import ballerina.file;

@http:configuration {port:9092}
service<http> multiparts {
    @http:resourceConfig {
        methods:["GET"],
        path:"/encode_out_response"
    }
    resource multipartSender (http:Connection conn, http:InRequest req) {

        //Create an enclosing entity to hold child parts.
        mime:Entity parentPart = {};
        mime:MediaType mixedContentType = mime:getMediaType(mime:MULTIPART_MIXED);
        parentPart.contentType = mixedContentType;

        //Create a child part with json content.
        mime:Entity childPart1 = {};
        mime:MediaType contentTypeOfJsonPart = mime:getMediaType(mime:APPLICATION_JSON);
        childPart1.contentType = contentTypeOfJsonPart;
        childPart1.setJson({"name":"wso2"});

        //Create another child part with a file.
        mime:Entity childPart2 = {};
        mime:MediaType contentTypeOfFilePart = mime:getMediaType(mime:TEXT_XML);
        childPart2.contentType = contentTypeOfFilePart;
        file:File fileHandler = {path:"/home/user/Downloads/test.xml"};
        childPart2.setFileAsEntityBody(fileHandler);

        //Create an array to hold child parts.
        mime:Entity[] childParts = [childPart1, childPart2];

        //Set the child parts to parent part.
        parentPart.setBodyParts(childParts);

        //Create an array to hold the parent part and set it to response.
        mime:Entity[] immediatePartsToResponse = [parentPart];
        http:OutResponse outResponse = {};
        outResponse.setMultiparts(immediatePartsToResponse, mime:MULTIPART_FORM_DATA);

        _ = conn.respond(outResponse);
    }
}
```

### CODE REVIEW

File: docs/ballerina-by-example/examples/inbound-response-with-multiparts/inbound-response-with-multiparts.bal [hunk 1]
```diff
@@ -0,0 +1,83 @@
+import ballerina.net.http;
```
Comment (Manuri) [inline docs/ballerina-by-example/examples/inbound-response-with-multiparts/inbound-response-with-multiparts.bal]: Missing license header

File: docs/ballerina-by-example/examples/outbound-response-with-multiparts/outbound-response-with-multiparts.bal [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import ballerina.net.http;
```
Comment (Manuri) [inline docs/ballerina-by-example/examples/outbound-response-with-multiparts/outbound-response-with-multiparts.bal]: Missing license header

File: stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/nativeimpl/connection/ConnectionAction.java [hunk 3]
```diff
@@ -61,15 +64,52 @@
...
+        BRefValueArray bodyParts = EntityBodyHandler.getBodyPartArray(entityStruct);
+        if (bodyParts != null && bodyParts.size() > 0) {
+            MultipartDataSource multipartDataSource = new MultipartDataSource(entityStruct, boundaryString);
+            serializeMsgDataSource(responseMessage, multipartDataSource, outboundRespStatusFuture,
+                    entityStruct);
+        } else {
+            OutputStream messageOutputStream = getOutputStream(responseMessage, outboundRespStatusFuture);
+            try {
+                EntityBodyHandler.writeByteChannelToOutputStream(entityStruct, messageOutputStream);
+                HttpUtil.closeMessageOutputStream(messageOutputStream);
```
Comment (Manuri) [inline stdlib/ballerina-http/src/main/java/org/ballerinalang/net/http/nativeimpl/connection/ConnectionAction.java]: If an exception occur before this, seems output stream would not be closed. Can we use finally clause here?

File: docs/ballerina-by-example/examples/inbound-response-with-multiparts/inbound-response-with-multiparts.bal (hunk 1 above)
Comment (arukshani) [reply docs/ballerina-by-example/examples/inbound-response-with-multiparts/inbound-response-with-multiparts.bal]: we do not add license headers to ballerina by samples

File: stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/response.bal [hunk 4]
```diff
@@ -158,6 +159,41 @@ public function <OutResponse response> getBinaryPayload () (blob) {
...
+@Description {value:"Gets the outbound response payload as a byte channel except for multiparts. In case of multiparts,
+please use 'getMultiparts()' instead."}
+@Param {value:"request: outbound response message"}
+@Return {value:"A byte channel as the message payload"}
+public function <OutResponse response> getByteChannel () (io:ByteChannel) {
+    mime:Entity entity = response.getEntity();
+    return entity.getByteChannel();
+}
+
+
```
Comment (anupama-pathirage) [inline stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/response.bal]: Extra new line?

File: stdlib/ballerina-mime/src/main/java/org/ballerinalang/mime/util/HeaderUtil.java [hunk 5]
```diff
@@ -248,4 +250,22 @@ static void setContentIdHeader(BStruct bodyPart, BMap<String, BValue> entityHead
...
     }
+
+    public static boolean isMultipart(String contentType) {
+        return contentType != null && contentType.startsWith(MULTIPART_AS_PRIMARY_TYPE);
+    }
+
+    /**
+     * Given a Content-Type, extract the boundary parameter value out of it.
+     *
+     * @param contentType Represent the value of Content-Type header inclusing parameters
```
Comment (anupama-pathirage) [inline stdlib/ballerina-mime/src/main/java/org/ballerinalang/mime/util/HeaderUtil.java]: Typo inclusing?

File: docs/ballerina-by-example/examples/inbound-response-with-multiparts/inbound-response-with-multiparts.bal (hunk 1 above)
Comment (Manuri) [reply docs/ballerina-by-example/examples/inbound-response-with-multiparts/inbound-response-with-multiparts.bal]: Oh ok! I didn't notice this is one of them.

File: stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/response.bal [hunk 6]
```diff
@@ -158,6 +159,40 @@ public function <OutResponse response> getBinaryPayload () (blob) {
     return entity.getBlob();
 }
 
+@Description {value:"Gets the inbound response payload as a byte channel except for multiparts. In case of multiparts,
+please use 'getMultiparts()' instead."}
+@Param {value:"request: The inbound response message"}
```
Comment (chamil321) [inline stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/response.bal]: variable name should be "response".

File: stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/response.bal [hunk 7]
```diff
@@ -158,6 +159,40 @@ public function <OutResponse response> getBinaryPayload () (blob) {
...
+@Return {value:"A byte channel as the message payload"}
+public function <InResponse response> getByteChannel () (io:ByteChannel) {
+    mime:Entity entity = response.getEntity();
+    return entity.getByteChannel();
+}
+
+@Description {value:"Gets the outbound response payload as a byte channel except for multiparts. In case of multiparts,
+please use 'getMultiparts()' instead."}
+@Param {value:"request: outbound response message"}
```
Comment (chamil321) [inline stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/response.bal]: variable name should be "response".

Comment (chamil321) [reply stdlib/ballerina-http/src/main/ballerina/ballerina/net/http/response.bal]: Please fix others similar occurrence

**Pull request merged**